streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableAggregate.java

```diff
@@ -116,10 +118,13 @@ public void process(final Record<KIn, Change<VIn>> record) {
             }
 
             // update the store with the new value
-            store.put(record.key(), newAgg, newTimestamp);
-            tupleForwarder.maybeForward(
-                record.withValue(new Change<>(newAgg, sendOldValues ? oldAgg : null))
-                    .withTimestamp(newTimestamp));
+            final long putReturnCode = store.put(record.key(), newAgg, newTimestamp);
+            // if not put to store, do not forward downstream either
+            if (putReturnCode != PUT_RETURN_CODE_NOT_PUT) {
```

Comment:
Some processors, including this one, guarantee that their timestamps are in non-decreasing order (per key), which means `putReturnCode` will always equal `PUT_RETURN_CODE_IS_LATEST`, so having these extra comparisons here is technically unnecessary, but it seems nice to have them anyway for consistency with other processors (and in case this guarantee ever changes, though it's unlikely). Happy to remove the redundancy if that's preferable, though.

Reply:
Cf my comments above -- not 100% sure if keeping the code is better or worse; tend to prefer to remove it and just add a comment about it?

Reply:
Is it valid for a computed aggregation to be null? If it is, I think it's actually not guaranteed that aggregation result timestamps are always nondecreasing. If an out-of-order record arrives after the `newAgg` value is set to null, then the aggregation result will be re-initialized, and the timestamp will be the timestamp of the out-of-order record, which is earlier than the previous aggregation timestamp.

If that scenario is possible, then it's a bit unclear which of the two should be considered the "latest" aggregation value, and we should chose whether to keep the current code or to always set `isLatest = true` depending on what semantics we want.

If that scenario is not possible, i.e., if it is not valid for an aggregation value to be null, then the two are equivalent and I can update this code to clean it up per your suggestion.